File: cpp/src/Ice/ServantManager.cpp
// Copyright (c) ZeroC, Inc.

#include "ServantManager.h"
#include "Ice/InputStream.h"
#include "Ice/LocalExceptions.h"
#include "Ice/LoggerUtil.h"
#include "Ice/ServantLocator.h"
#include "Ice/StringUtil.h"
#include "Instance.h"

using namespace std;
using namespace Ice;
using namespace IceInternal;

Ice::ServantLocator::~ServantLocator() = default; // avoid weak vtable

void
IceInternal::ServantManager::addServant(ObjectPtr object, Identity ident, string facet)
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;

    if (p == _servantMapMap.end() || p->first != ident)
    {
        p = _servantMapMap.find(ident);
    }

    if (p == _servantMapMap.end())
    {
        p = _servantMapMap.insert(_servantMapMapHint, pair<const Identity, FacetMap>(std::move(ident), FacetMap()));
    }
    else
    {
        if (p->second.find(facet) != p->second.end())
        {
            ToStringMode toStringMode = _instance->toStringMode();
            ostringstream os;
            os << Ice::identityToString(ident, toStringMode);
            if (!facet.empty())
            {
                os << " -f " << escapeString(facet, "", toStringMode);
            }
            throw AlreadyRegisteredException(__FILE__, __LINE__, "servant", os.str());
        }
    }

    _servantMapMapHint = p;

    p->second.insert(pair<const string, ObjectPtr>(std::move(facet), std::move(object)));
}

void
IceInternal::ServantManager::addDefaultServant(ObjectPtr object, string category)
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _defaultServantMap.find(category);
    if (p != _defaultServantMap.end())
    {
        throw AlreadyRegisteredException(__FILE__, __LINE__, "default servant", category);
    }

    _defaultServantMap.insert(pair<const string, ObjectPtr>(std::move(category), std::move(object)));
}

ObjectPtr
IceInternal::ServantManager::removeServant(const Identity& ident, const string_view facet)
{
    // We return the removed servant to avoid releasing the last reference count
    // with *this locked. We don't want to run user code, such as the servant
    // destructor, with an internal Ice mutex locked.

    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;
    FacetMap::iterator q;

    if (p == _servantMapMap.end() || p->first != ident)
    {
        p = _servantMapMap.find(ident);
    }

    if (p == _servantMapMap.end() || (q = p->second.find(facet)) == p->second.end())
    {
        ToStringMode toStringMode = _instance->toStringMode();
        ostringstream os;
        os << Ice::identityToString(ident, toStringMode);
        if (!facet.empty())
        {
            os << " -f " + escapeString(facet, "", toStringMode);
        }
        throw NotRegisteredException(__FILE__, __LINE__, "servant", os.str());
    }

    ObjectPtr servant = q->second;
    p->second.erase(q);

    if (p->second.empty())
    {
        if (p == _servantMapMapHint)
        {
            _servantMapMap.erase(p++);
            _servantMapMapHint = p;
        }
        else
        {
            _servantMapMap.erase(p);
        }
    }
    return servant;
}

ObjectPtr
IceInternal::ServantManager::removeDefaultServant(const string_view category)
{
    // We return the removed servant to avoid releasing the last reference count
    // with *this locked. We don't want to run user code, such as the servant
    // destructor, with an internal Ice mutex locked.

    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _defaultServantMap.find(category);
    if (p == _defaultServantMap.end())
    {
        throw NotRegisteredException(__FILE__, __LINE__, "default servant", string{category});
    }

    ObjectPtr servant = p->second;
    _defaultServantMap.erase(p);

    return servant;
}

FacetMap
IceInternal::ServantManager::removeAllFacets(const Identity& ident)
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;

    if (p == _servantMapMap.end() || p->first != ident)
    {
        p = _servantMapMap.find(ident);
    }

    if (p == _servantMapMap.end())
    {
        throw NotRegisteredException(
            __FILE__,
            __LINE__,
            "servant",
            Ice::identityToString(ident, _instance->toStringMode()));
    }

    FacetMap result = p->second;

    if (p == _servantMapMapHint)
    {
        _servantMapMap.erase(p++);
        _servantMapMapHint = p;
    }
    else
    {
        _servantMapMap.erase(p);
    }

    return result;
}

ObjectPtr
IceInternal::ServantManager::findServant(const Identity& ident, const string_view facet) const
{
    lock_guard lock(_mutex);

    //
    // This assert is not valid if the adapter dispatch incoming
    // requests from bidir connections. This method might be called if
    // requests are received over the bidir connection after the
    // adapter was deactivated.
    //
    // assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;
    FacetMap::iterator q;

    auto& servantMapMap = const_cast<ServantMapMap&>(_servantMapMap);

    if (p == servantMapMap.end() || p->first != ident)
    {
        p = servantMapMap.find(ident);
    }

    if (p == servantMapMap.end() || (q = p->second.find(facet)) == p->second.end())
    {
        auto d = _defaultServantMap.find(ident.category);
        if (d == _defaultServantMap.end())
        {
            d = _defaultServantMap.find("");
            if (d == _defaultServantMap.end())
            {
                return nullptr;
            }
            else
            {
                return d->second;
            }
        }
        else
        {
            return d->second;
        }
    }
    else
    {
        _servantMapMapHint = p;
        return q->second;
    }
}

ObjectPtr
IceInternal::ServantManager::findDefaultServant(const string_view category) const
{
    lock_guard lock(_mutex);

    auto p = _defaultServantMap.find(category);
    if (p == _defaultServantMap.end())
    {
        return nullptr;
    }
    else
    {
        return p->second;
    }
}

FacetMap
IceInternal::ServantManager::findAllFacets(const Identity& ident) const
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;

    auto& servantMapMap = const_cast<ServantMapMap&>(_servantMapMap);

    if (p == servantMapMap.end() || p->first != ident)
    {
        p = servantMapMap.find(ident);
    }

    if (p == servantMapMap.end())
    {
        return {};
    }
    else
    {
        _servantMapMapHint = p;
        return p->second;
    }
}

bool
IceInternal::ServantManager::hasServant(const Identity& ident) const
{
    lock_guard lock(_mutex);

    //
    // This assert is not valid if the adapter dispatch incoming
    // requests from bidir connections. This method might be called if
    // requests are received over the bidir connection after the
    // adapter was deactivated.
    //
    // assert(_instance); // Must not be called after destruction.

    auto p = _servantMapMapHint;
    auto& servantMapMap = const_cast<ServantMapMap&>(_servantMapMap);

    if (p == servantMapMap.end() || p->first != ident)
    {
        p = servantMapMap.find(ident);
    }

    if (p == servantMapMap.end())
    {
        return false;
    }
    else
    {
        _servantMapMapHint = p;
        assert(!p->second.empty());
        return true;
    }
}

void
IceInternal::ServantManager::addServantLocator(ServantLocatorPtr locator, string category)
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    if ((_locatorMapHint != _locatorMap.end() && _locatorMapHint->first == category) ||
        _locatorMap.find(category) != _locatorMap.end())
    {
        throw AlreadyRegisteredException(__FILE__, __LINE__, "servant locator", category);
    }

    _locatorMapHint = _locatorMap.insert(
        _locatorMapHint,
        pair<const string, ServantLocatorPtr>(std::move(category), std::move(locator)));
}

ServantLocatorPtr
IceInternal::ServantManager::removeServantLocator(const string_view category)
{
    lock_guard lock(_mutex);

    assert(_instance); // Must not be called after destruction.

    auto p = _locatorMap.end();
    if (_locatorMapHint != p)
    {
        if (_locatorMapHint->first == category)
        {
            p = _locatorMapHint;
        }
    }

    if (p == _locatorMap.end())
    {
        p = _locatorMap.find(category);
    }

    if (p == _locatorMap.end())
    {
        throw NotRegisteredException(__FILE__, __LINE__, "servant locator", string{category});
    }

    ServantLocatorPtr locator = p->second;
    _locatorMap.erase(p);
    _locatorMapHint = _locatorMap.begin();
    return locator;
}

ServantLocatorPtr
IceInternal::ServantManager::findServantLocator(const string_view category) const
{
    lock_guard lock(_mutex);

    //
    // This assert is not valid if the adapter dispatch incoming
    // requests from bidir connections. This method might be called if
    // requests are received over the bidir connection after the
    // adapter was deactivated.
    //
    // assert(_instance); // Must not be called after destruction.

    auto& locatorMap = const_cast<map<string, ServantLocatorPtr, std::less<>>&>(_locatorMap);

    auto p = locatorMap.end();
    if (_locatorMapHint != locatorMap.end())
    {
        if (_locatorMapHint->first == category)
        {
            p = _locatorMapHint;
        }
    }

    if (p == locatorMap.end())
    {
        p = locatorMap.find(category);
    }

    if (p != locatorMap.end())
    {
        _locatorMapHint = p;
        return p->second;
    }
    else
    {
        return nullptr;
    }
}

IceInternal::ServantManager::ServantManager(InstancePtr instance, string adapterName)
    : _instance(std::move(instance)),
      _adapterName(std::move(adapterName)),
      _servantMapMapHint(_servantMapMap.end()),
      _locatorMapHint(_locatorMap.end())
{
}

IceInternal::ServantManager::~ServantManager() = default;

void
IceInternal::ServantManager::destroy()
{
    ServantMapMap servantMapMap;
    DefaultServantMap defaultServantMap;
    map<string, ServantLocatorPtr, std::less<>> locatorMap;
    Ice::LoggerPtr logger;

    {
        lock_guard lock(_mutex);
        //
        // If the ServantManager has already been destroyed, we're done.
        //
        if (!_instance)
        {
            return;
        }

        logger = _instance->initializationData().logger;

        servantMapMap.swap(_servantMapMap);
        _servantMapMapHint = _servantMapMap.end();

        defaultServantMap.swap(_defaultServantMap);

        locatorMap.swap(_locatorMap);
        _locatorMapHint = _locatorMap.end();

        _instance = nullptr;
    }

    for (const auto& p : locatorMap)
    {
        try
        {
            p.second->deactivate(p.first);
        }
        catch (const Exception& ex)
        {
            Error out(logger);
            out << "exception during locator deactivation:\n"
                << "object adapter: '" << _adapterName << "'\n"
                << "locator category: '" << p.first << "'\n"
                << ex;
        }
        catch (...)
        {
            Error out(logger);
            out << "unknown exception during locator deactivation:\n"
                << "object adapter: '" << _adapterName << "'\n"
                << "locator category: '" << p.first << "'";
        }
    }

    //
    // We clear the maps outside the synchronization as we don't want to
    // hold any internal Ice mutex while running user code (such as servant
    // or servant locator destructors).
    //
    servantMapMap.clear();
    locatorMap.clear();
    defaultServantMap.clear();
}

void
ServantManager::dispatch(IncomingRequest& request, function<void(OutgoingResponse)> sendResponse)
{
    const Current& current = request.current();
    ObjectPtr servant = findServant(current.id, current.facet);

    if (servant)
    {
        // the simple, common path.
        servant->dispatch(request, std::move(sendResponse));
        return;
    }

    // Else, check servant locators
    ServantLocatorPtr locator = findServantLocator(current.id.category);
    if (!locator && !current.id.category.empty())
    {
        locator = findServantLocator("");
    }

    if (locator)
    {
        shared_ptr<void> cookie;
        try
        {
            servant = locator->locate(current, cookie);
        }
        catch (...)
        {
            // Skip the encapsulation. This allows the next batch requests in the same InputStream to proceed.
            request.inputStream().skipEncapsulation();
            throw;
        }

        if (servant)
        {
            try
            {
                servant->dispatch(
                    request,
                    [sendResponse = std::move(sendResponse), locator, servant, cookie](OutgoingResponse response)
                    {
                        try
                        {
                            locator->finished(response.current(), servant, cookie);
                        }
                        catch (...)
                        {
                            sendResponse(makeOutgoingResponse(std::current_exception(), response.current()));
                            return; // done
                        }
                        sendResponse(std::move(response));
                    });
                return; // done
            }
            catch (...)
            {
                // When we catch an exception, the dispatch logic guarantees sendResponse was not called.
                locator->finished(current, servant, cookie);
                throw;
            }
        }
    }

    assert(!servant);

    // Skip the encapsulation. This allows the next batch requests in the same InputStream to proceed.
    request.inputStream().skipEncapsulation();
    if (hasServant(current.id))
    {
        throw FacetNotExistException{__FILE__, __LINE__};
    }
    else
    {
        throw ObjectNotExistException{__FILE__, __LINE__};
    }
}

File: js/src/Ice/ServantManager.js
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import { HashMap } from "./HashMap.js";
import { StringUtil } from "./StringUtil.js";
import {
    AlreadyRegisteredException,
    NotRegisteredException,
    ObjectNotExistException,
    FacetNotExistException,
} from "./LocalExceptions.js";
import { identityToString } from "./IdentityToString.js";
import { Debug } from "./Debug.js";

//
// Only for use by Ice.ObjectAdapter.
//
export class ServantManager {
    constructor(instance, adapterName) {
        this._instance = instance;
        this._adapterName = adapterName;
        // Map<Ice.Identity, Map<String, Ice.Object> >
        this._servantMapMap = new HashMap(HashMap.compareEquals);
        // Map<String, Ice.Object>
        this._defaultServantMap = new Map();
        // Map<String, Ice.ServantLocator>
        this._locatorMap = new Map();
    }

    async dispatch(request) {
        const current = request.current;
        let servant = this.findServant(current.id, current.facet);

        if (servant !== null) {
            // the simple, common path
            return await servant.dispatch(request);
        }

        // Else, check servant locators
        let locator = this.findServantLocator(current.id.category);
        if (locator === null && current.id.category.length > 0) {
            locator = this.findServantLocator("");
        }

        if (locator !== null) {
            const cookie = {};
            try {
                servant = locator.locate(current, cookie);
            } catch (ex) {
                // Skip the encapsulation. This allows the next batch requests in the same InputStream to proceed.
                request.inputStream.skipEncapsulation();
                throw ex;
            }

            if (servant !== null) {
                try {
                    return await servant.dispatch(request);
                } finally {
                    locator.finished(current, servant, cookie.value);
                }
            }
        }

        DEV: console.assert(servant === null);

        // Skip the encapsulation. This allows the next batch requests in the same InputStream to proceed.
        request.inputStream.skipEncapsulation();
        if (this.hasServant(current.id)) {
            throw new FacetNotExistException();
        } else {
            throw new ObjectNotExistException();
        }
    }

    addServant(servant, ident, facet) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        if (facet === null) {
            facet = "";
        }

        let m = this._servantMapMap.get(ident);
        if (m === undefined) {
            m = new Map();
            this._servantMapMap.set(ident, m);
        } else if (m.has(facet)) {
            let id = identityToString(ident, this._instance.toStringMode());
            if (facet.length > 0) {
                id += " -f " + StringUtil.escapeString(facet, "", this._instance.toStringMode());
            }
            throw new AlreadyRegisteredException("servant", id);
        }

        m.set(facet, servant);
    }

    addDefaultServant(servant, category) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction

        if (this._defaultServantMap.has(category)) {
            throw new AlreadyRegisteredException("default servant", category);
        }

        this._defaultServantMap.set(category, servant);
    }

    removeServant(ident, facet) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        if (facet === null) {
            facet = "";
        }

        const m = this._servantMapMap.get(ident);
        if (m === undefined || !m.has(facet)) {
            let id = identityToString(ident, this._instance.toStringMode());
            if (facet.length > 0) {
                id += " -f " + StringUtil.escapeString(facet, "", this._instance.toStringMode());
            }
            throw new NotRegisteredException("servant", id);
        }

        const obj = m.get(facet);
        m.delete(facet);

        if (m.size === 0) {
            this._servantMapMap.delete(ident);
        }

        return obj;
    }

    removeDefaultServant(category) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        const obj = this._defaultServantMap.get(category);
        if (obj === undefined) {
            throw new NotRegisteredException("default servant", category);
        }

        this._defaultServantMap.delete(category);
        return obj;
    }

    removeAllFacets(ident) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        const m = this._servantMapMap.get(ident);
        if (m === undefined) {
            throw new NotRegisteredException("servant", identityToString(ident, this._instance.toStringMode()));
        }

        this._servantMapMap.delete(ident);

        return m;
    }

    findServant(ident, facet) {
        if (facet === null) {
            facet = "";
        }

        const m = this._servantMapMap.get(ident);
        let obj = null;
        if (m === undefined) {
            obj = this._defaultServantMap.get(ident.category);
            if (obj === undefined) {
                obj = this._defaultServantMap.get("");
            }
        } else {
            obj = m.get(facet);
        }

        return obj === undefined ? null : obj;
    }

    findDefaultServant(category) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        const ds = this._defaultServantMap.get(category);
        return ds === undefined ? null : ds;
    }

    findAllFacets(ident) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        const m = this._servantMapMap.get(ident);
        if (m !== undefined) {
            return new Map(m);
        }

        return new Map();
    }

    hasServant(ident) {
        const m = this._servantMapMap.get(ident);
        if (m === undefined) {
            return false;
        } else {
            DEV: console.assert(m.size > 0);
            return true;
        }
    }

    addServantLocator(locator, category) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        if (this._locatorMap.has(category)) {
            throw new AlreadyRegisteredException(
                "servant locator",
                StringUtil.escapeString(category, "", this._instance.toStringMode()),
            );
        }

        this._locatorMap.set(category, locator);
    }

    removeServantLocator(category) {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.

        const l = this._locatorMap.get(category);
        if (l === undefined) {
            throw new NotRegisteredException(
                "servant locator",
                StringUtil.escapeString(category, "", this._instance.toStringMode()),
            );
        }
        this._locatorMap.delete(category);
        return l;
    }

    findServantLocator(category) {
        const l = this._locatorMap.get(category);
        return l === undefined ? null : l;
    }

    //
    // Only for use by Ice.ObjectAdapterI.
    //
    destroy() {
        DEV: console.assert(this._instance !== null); // Must not be called after destruction.
        const logger = this._instance.initializationData().logger;
        this._servantMapMap.clear();

        this._defaultServantMap.clear();

        const locatorMap = new Map(this._locatorMap);
        this._locatorMap.clear();
        this._instance = null;

        for (const [key, locator] of locatorMap) {
            try {
                locator.deactivate(key);
            } catch (ex) {
                logger.error(
                    "exception during locator deactivation:\nobject adapter: `" +
                        this._adapterName +
                        "'\nlocator category: `" +
                        key +
                        "'\n" +
                        ex.toString(),
                );
            }
        }
    }
}
